sql/core/src/main/scala/org/apache/spark/sql/streaming/TwsTester.scala

```diff
@@ -0,0 +1,286 @@
+ *  - TTL for ValueState, ListState, and MapState (use ProcessingTime mode and
+ *    `advanceProcessingTime` to test expiry).
+ *
+ * '''Testing EventTime Mode:'''
+ *  To test with EventTime, provide `eventTimeExtractor` (a function extracting the event
+ *  timestamp from each input row) and `watermarkDelayMs` (the watermark delay in milliseconds).
+ *  The watermark is computed as `max(event_time_seen) - watermarkDelayMs` and is updated
+ *  automatically after each `test()` call. Late events (with event time older than the current
+ *  watermark) are filtered out before processing, matching production behavior. Timers with
+ *  expiry time <= watermark will fire. You can also manually advance the watermark using
+ *  `advanceWatermark()`.
+ *
+ * '''Use Cases:'''
+ *  - '''Primary''': Unit testing business logic in `handleInputRows` implementations.
+ *  - '''Not recommended''': End-to-end testing or performance testing - use actual Spark
+ *    streaming queries for those scenarios.
+ *
+ * @param processor the StatefulProcessor to test.
+ * @param initialState initial state for each key as a list of (key, state) tuples.
+ * @param timeMode time mode (None, ProcessingTime or EventTime).
+ * @param outputMode output mode (Append, Update, or Complete).
+ * @param realTimeMode whether input rows should be processed one-by-one (separate call to
+ *     handleInputRows) for each input row.
+ * @param eventTimeExtractor function to extract event time from input rows. Required if and
+ *     only if timeMode is EventTime.
+ * @param watermarkDelayMs watermark delay in milliseconds. The watermark is computed as
+ *     `max(event_time) - watermarkDelayMs`. Required if and only if timeMode is EventTime.
+ * @tparam K the type of grouping key.
+ * @tparam I the type of input rows.
+ * @tparam O the type of output rows.
+ * @since 4.0.2
+ */
+class TwsTester[K, I, O](
+    val processor: StatefulProcessor[K, I, O],
+    val initialState: List[(K, Any)] = List(),
+    val timeMode: TimeMode = TimeMode.None,
+    val outputMode: OutputMode = OutputMode.Append,
+    val realTimeMode: Boolean = false,
+    val eventTimeExtractor: I => Timestamp = null,
+    val watermarkDelayMs: Long = 0L) {
+  val clock: Clock = new Clock {
+    override def instant(): Instant = Instant.ofEpochMilli(currentProcessingTimeMs)
+    override def getZone: ZoneId = ZoneId.systemDefault()
+    override def withZone(zone: ZoneId): Clock = this
+  }
+
+  private val handle = new InMemoryStatefulProcessorHandle(timeMode, clock)
+
+  if (timeMode == TimeMode.EventTime) {
+    require(
+      eventTimeExtractor != null,
+      "eventTimeExtractor must be provided when timeMode is EventTime."
+    )
+  }
+
+  processor.setHandle(handle)
+  processor.init(outputMode, timeMode)
+  processor match {
+    case p: StatefulProcessorWithInitialState[K @unchecked, I @unchecked, O @unchecked, s] =>
```

Comment:
Good to know the type can be captured and be passed into the type info again. TIL.